### PR TITLE
Upgrade from Chromium 109 to Chromium 110

### PR DESCRIPTION
**Pull request by emerick**
Resolves https://github.com/brave/brave-browser/issues/27221

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by Brandon-T — APPROVED** [review]

**Review by kylehickinson — APPROVED** [review]
iOS++

**Review by zenparsing — APPROVED** [review]
Rewards 👍

**Review by rillian — APPROVED** [review]

**Review by fmarier — APPROVED** [review]
`chromium_src/net/tools/transport_security_state_generator/input_file_parsers.cc` changes look good.

**Review by tmancey — APPROVED** [review]
ads LGTM

**Review by wchen342 — APPROVED** [review]
Android tests++

**Review by darkdh — APPROVED** [review]

**Review by simonhong — APPROVED** [review]

File: chromium_src/components/sync/base/model_type_unittest.cc [hunk 1]
```diff
@@ -14,7 +14,7 @@ TEST_F(ModelTypeTest, EncryptableUserTypes) {
 
 // This test is supposed to fail when sync types are increased/decreased
 TEST_F(ModelTypeTest, ModelTypeCounts) {
-  EXPECT_EQ(static_cast<int>(ModelTypeForHistograms::kMaxValue), 56);
+  EXPECT_EQ(static_cast<int>(ModelTypeForHistograms::kMaxValue), 57);
```
Comment (AlexeyBarabash) [inline chromium_src/components/sync/base/model_type_unittest.cc:17]: New datatype is `kPowerBookmark`.
https://source.chromium.org/chromium/chromium/src/+/main:components/power_bookmarks/core/proto/power_bookmark_meta.proto;bpv=1;bpt=1

Comment (AlexeyBarabash) [reply chromium_src/components/sync/base/model_type_unittest.cc:17]: https://github.com/brave/brave-browser/issues/27928

**Review by AlexeyBarabash — APPROVED** [review]
Sync changes ++, simple sync test Android <=> Desktop works well

File: chromium_src/chrome/browser/ui/views/bubble/bubble_contents_wrapper.h [hunk 2]
```diff
@@ -24,15 +24,17 @@
                       WindowOpenDisposition disposition,                   \
                       const blink::mojom::WindowFeatures& window_features, \
                       bool user_gesture, bool* was_blocked) override;      \
-  std::vector<int32_t>& GetPopupIds() { return popup_ids_; }               \
-  void RenderViewHostChanged
+  std::vector<int32_t>& GetPopupIds() {                                    \
```
Comment (goodov) [inline chromium_src/chrome/browser/ui/views/bubble/bubble_contents_wrapper.h:27]: nit: 
```
const std::vector<int32_t>& popup_ids() const { return popup_ids_; }
void ClearPopupIds();

.cc:
ClearPopupIds() { popup_ids.clear(); }
```

File: chromium_src/components/permissions/permission_context_base.cc [hunk 3]
```diff
@@ -63,9 +63,32 @@ void PermissionContextBase::PermissionDecided(const PermissionRequestID& id,
...
+  UserMadePermissionDecision(id, requesting_origin, embedding_origin,
+                             content_setting);
+
+  bool persist = content_setting != CONTENT_SETTING_DEFAULT;
+
+  auto grouped_request = pending_grouped_requests_.find(id.ToString());
+  DCHECK(grouped_request != pending_grouped_requests_.end());
+  DCHECK(grouped_request->second);
+
+  auto callback = grouped_request->second->GetNextCallback();
```
Comment (goodov) [inline chromium_src/components/permissions/permission_context_base.cc:86]: shouldn't we check `!IsDone()` here? Can you please describe what exactly this change fixes? Maybe extend a commit message to include more details.

File: test/BUILD.gn [hunk 4]
```diff
@@ -183,7 +183,6 @@ test("brave_unit_tests") {
     "//brave/browser/ui/brave_tooltips",
     "//brave/browser/ui/webui/new_tab_page:unit_tests",
     "//brave/browser/web_discovery:unit_tests",
-    "//brave/browser/widevine:unittest",
```
Comment (goodov) [inline test/BUILD.gn]: that removal is valid. We have a correct dep that uses `if (enable_widevine)` below:
https://github.com/brave/brave-core/blob/437a66e377eb02b132b36e51bcd3a07b2a258569/test/BUILD.gn#L315-L317

File: patches/components-crash-core-browser-resources-crashes.ts.patch [hunk 5]
```diff
@@ -0,0 +1,13 @@
+diff --git a/components/crash/core/browser/resources/crashes.ts b/components/crash/core/browser/resources/crashes.ts
+index 4c7d88670a8c00552b126e3a4cc6c28b26d002c7..8a882c542af26ae686c9b06b84a4968e1887fc7d 100644
+--- a/components/crash/core/browser/resources/crashes.ts
++++ b/components/crash/core/browser/resources/crashes.ts
+@@ -141,7 +141,7 @@ function updateCrashList({
+         crashLink.href = `https://goto.google.com/crash/${crash.id}`;
+         crashLink.target = '_blank';
+         crashLink.textContent = crash.id;
+-        uploadIdValue.appendChild(crashLink);
++        // uploadIdValue.appendChild(crashLink);
```
[the remark this answers is not in the document]
Comment (emerick) [reply patches/components-crash-core-browser-resources-crashes.ts.patch]: Fixed in https://github.com/brave/brave-core/pull/16324/commits/6dd7ad2f59823d4dc0ad153b9e5b2be56566b148

File: chromium_src/components/update_client/update_checker.cc [hunk 6]
```diff
@@ -5,6 +5,8 @@
 
 #include "src/components/update_client/update_checker.cc"
 
+#include "base/threading/thread_task_runner_handle.h"
```
[the remark this answers is not in the document]
Comment (emerick) [reply chromium_src/components/update_client/update_checker.cc]: Thanks, fixed in https://github.com/brave/brave-core/pull/16324/commits/87c25c3dd247faf2d32fde8978c6e0c4105639f0

File: browser/ui/views/tabs/brave_browser_tab_strip_controller.cc [hunk 7]
```diff
@@ -31,8 +31,6 @@ void BraveBrowserTabStripController::ShowContextMenuForTab(
   BrowserView* browser_view =
       BrowserView::GetBrowserViewForBrowser(browser());
   context_menu_contents_ = std::make_unique<BraveTabContextMenuContents>(
-      tab,
-      this,
-      browser_view->tabstrip()->GetModelIndexOf(tab));
+      tab, this, *browser_view->tabstrip()->GetModelIndexOf(tab));
```
[the remark this answers is not in the document]
Comment (emerick) [reply browser/ui/views/tabs/brave_browser_tab_strip_controller.cc]: Good idea, fixed in https://github.com/brave/brave-core/pull/16324/commits/30e5f3a25e2dfbe315d7d11d3de685e42802c268.

File: net/http/partitioned_host_state_map.cc [hunk 8]
```diff
@@ -7,46 +7,58 @@
...
-#include "base/strings/strcat.h"
+#include "base/ranges/algorithm.h"
 #include "crypto/sha2.h"
 #include "net/base/network_isolation_key.h"
 
+namespace {
+
+bool IsEmptyPartitionHash(
+    const net::PartitionedHostStateMapBase::HashedHost& hashed_host) {
+  return std::all_of(hashed_host.begin(), hashed_host.end(),
```
[the remark this answers is not in the document]
Comment (emerick) [reply net/http/partitioned_host_state_map.cc]: Fixed in https://github.com/brave/brave-core/pull/16324/commits/4d2b035e8992ff9ded6c8174dde2a5665edf297e.

File: net/http/partitioned_host_state_map.cc [hunk 9]
```diff
@@ -7,46 +7,58 @@
...
+PartitionedHostStateMapBase::HashedHost
+PartitionedHostStateMapBase::GetKeyWithPartitionHash(
+    const HashedHost& k) const {
   CHECK(IsPartitionHashValid());
   if (k == *partition_hash_) {
     return k;
   }
-  return base::StrCat({GetHalfKey(k), GetHalfKey(*partition_hash_)});
+
+  std::array<uint8_t, crypto::kSHA256Length> result;
```
[the remark this answers is not in the document]
Comment (emerick) [reply net/http/partitioned_host_state_map.cc]: Fixed in https://github.com/brave/brave-core/pull/16324/commits/4d2b035e8992ff9ded6c8174dde2a5665edf297e.

File: net/http/partitioned_host_state_map.cc [hunk 10]
```diff
@@ -7,46 +7,58 @@
...
 }
 
-// static
-base::StringPiece PartitionedHostStateMapBase::GetHalfKey(base::StringPiece k) {
-  CHECK_EQ(k.size(), crypto::kSHA256Length);
-  const size_t kHalfSHA256HashLength = crypto::kSHA256Length / 2;
-  return k.substr(0, kHalfSHA256HashLength);
+base::span<const uint8_t> PartitionedHostStateMapBase::GetHalfKey(
+    const HashedHost& k) {
+  return base::make_span(k.data(), crypto::kSHA256Length / 2);
```
[the remark this answers is not in the document]
Comment (emerick) [reply net/http/partitioned_host_state_map.cc]: Fixed in https://github.com/brave/brave-core/pull/16324/commits/4d2b035e8992ff9ded6c8174dde2a5665edf297e.

File: net/http/partitioned_host_state_map.cc [hunk 11]
```diff
@@ -7,46 +7,58 @@
...
   CHECK(IsPartitionHashValid());
   if (k == *partition_hash_) {
     return k;
   }
-  return base::StrCat({GetHalfKey(k), GetHalfKey(*partition_hash_)});
+
+  std::array<uint8_t, crypto::kSHA256Length> result;
+  base::ranges::copy(GetHalfKey(k), result.begin());
+  base::ranges::copy(GetHalfKey(*partition_hash_),
+                     std::next(result.begin(), crypto::kSHA256Length / 2));
```
[the remark this answers is not in the document]
Comment (emerick) [reply net/http/partitioned_host_state_map.cc]: Fixed in https://github.com/brave/brave-core/pull/16324/commits/4d2b035e8992ff9ded6c8174dde2a5665edf297e.

File: net/http/partitioned_host_state_map.h [hunk 12]
```diff
@@ -27,26 +31,27 @@ class NET_EXPORT PartitionedHostStateMapBase {
...
   bool HasPartitionHash() const;
   // Returns true if |partition_hash_| contains a non empty valid hash.
   bool IsPartitionHashValid() const;
   // Creates a host hash by concatenating first 16 bytes (half of SHA256) from
   // |k| and first 16 bytes from |partition_hash_|.
   // CHECKs if |partition_hash_| is not valid.
-  std::string GetKeyWithPartitionHash(const std::string& k) const;
+  HashedHost GetKeyWithPartitionHash(const HashedHost& k) const;
 
   // Returns first 16 bytes from |k|.
```
[the remark this answers is not in the document]
Comment (emerick) [reply net/http/partitioned_host_state_map.h:44]: Fixed in https://github.com/brave/brave-core/pull/16324/commits/4d2b035e8992ff9ded6c8174dde2a5665edf297e.

File: net/http/partitioned_host_state_map.h [hunk 13]
```diff
@@ -111,15 +116,23 @@ class NET_EXPORT PartitionedHostStateMap : public PartitionedHostStateMapBase {
   // Removes all items with similar first 16 bytes of |k|, effectively ignoring
   // partition hash part.
   bool DeleteDataInAllPartitions(const key_type& k) {
-    auto equal_range_pair = base::ranges::equal_range(
-        map_, GetHalfKey(k), {},
-        [](const value_type& v) { return GetHalfKey(v.first); });
```
[the remark this answers is not in the document]
Comment (emerick) [reply net/http/partitioned_host_state_map.h:116]: Fixed in https://github.com/brave/brave-core/pull/16324/commits/4d2b035e8992ff9ded6c8174dde2a5665edf297e.

File: net/http/partitioned_host_state_map_unittest.cc [hunk 14]
```diff
@@ -8,20 +8,24 @@
 #include <map>
 #include <string>
 
+#include "base/containers/span.h"
 #include "crypto/sha2.h"
 #include "testing/gtest/include/gtest/gtest.h"
 
 namespace net {
 
 namespace {
```
[the remark this answers is not in the document]
Comment (emerick) [reply net/http/partitioned_host_state_map_unittest.cc:18]: Fixed in https://github.com/brave/brave-core/pull/16324/commits/4d2b035e8992ff9ded6c8174dde2a5665edf297e.

File: chromium_src/chrome/browser/ui/views/bubble/bubble_contents_wrapper.h (hunk 2 above)
Comment (emerick) [reply chromium_src/chrome/browser/ui/views/bubble/bubble_contents_wrapper.h]: https://github.com/brave/brave-core/pull/16324/commits/47df15e495536f7986f175b5b6a85cd250fea278

File: chromium_src/components/permissions/permission_context_base.cc (hunk 3 above)
Comment (emerick) [reply chromium_src/components/permissions/permission_context_base.cc:90]: Added that check and fleshed out the commit message to explain things better here https://github.com/brave/brave-core/pull/16324/commits/279574fdd6180c9ec2831259a2c77e24ca4e9917.

@yrliou May be able to provide additional context here?

**Review by bsclifton — APPROVED** [review]
JS deps changes LGTM!

File: browser/brave_stats/brave_stats_updater.cc [hunk 15]
```diff
@@ -117,7 +117,7 @@ BraveStatsUpdater::BraveStatsUpdater(PrefService* pref_service)
 BraveStatsUpdater::~BraveStatsUpdater() = default;
 
 void BraveStatsUpdater::OnProfileAdded(Profile* profile) {
-  if (profile == ProfileManager::GetPrimaryUserProfile()) {
+  if (profile == ProfileManager::GetLastUsedProfile()) {
```
Comment (bridiver) [inline browser/brave_stats/brave_stats_updater.cc:120]: @DJAndries

File: chromium_src/extensions/browser/sandboxed_unpacker.cc [hunk 16]
```diff
@@ -41,10 +41,8 @@ bool BraveRewriteManifest(const std::string& extension_id,
 
 }  // namespace brave
 
-#define BRAVE_SANDBOXEDUNPACKER_REWRITEMANIFESTFILE              \
-  base::Value::Dict* dict_manifest = final_manifest.GetIfDict(); \
-  DCHECK(dict_manifest);                                         \
-  brave::BraveRewriteManifest(extension_id_, dict_manifest);
+#define BRAVE_SANDBOXEDUNPACKER_REWRITEMANIFESTFILE \
+  brave::BraveRewriteManifest(extension_id_, &final_manifest);
```
Comment (bridiver) [inline chromium_src/extensions/browser/sandboxed_unpacker.cc:45]: is this even relevant anymore?

File: patches/chrome-browser-buildflags.gni.patch [hunk 17]
```diff
@@ -0,0 +1,16 @@
+diff --git a/chrome/browser/buildflags.gni b/chrome/browser/buildflags.gni
+index 653d675e736d55330ef86a62436f1d0c16406e60..fb014f4ea6882759cb2f78e11ece27fd01c60cf9 100644
+--- a/chrome/browser/buildflags.gni
++++ b/chrome/browser/buildflags.gni
+@@ -6,9 +6,9 @@ import("//build/config/chrome_build.gni")
+ 
+ declare_args() {
+   # By default, only branded builds integrate with automatic updates.
+-  enable_updater = is_chrome_branded
++  enable_updater = is_chrome_branded || is_official_build
```
Comment (bridiver) [inline patches/chrome-browser-buildflags.gni.patch:10]: this is a declare_args section so we can set this in config.js

File: patches/chrome-browser-buildflags.gni.patch [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+ 
+   # Detect updates and notify the user for Google Chrome across all platforms.
+   # Chromium does not use an auto-updater.
+-  enable_update_notifications = is_chrome_branded
++  enable_update_notifications = is_chrome_branded || is_official_build
```
Comment (bridiver) [inline patches/chrome-browser-buildflags.gni.patch:15]: same as above

File: browser/brave_stats/brave_stats_updater.cc (hunk 15 above)
Comment (DJAndries) [reply browser/brave_stats/brave_stats_updater.cc:120]: Should disappear after rebasing

**Review by muliswilliam — APPROVED** [review]

File: net/http/partitioned_host_state_map.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (goodov) [reply net/http/partitioned_host_state_map.h:116]: `return memcmp(v1.data(), v2.data(), v1.size()) < 0;`
good call! I missed that.

File: net/http/partitioned_host_state_map.h [hunk 19]
```diff
@@ -37,8 +37,7 @@ class NET_EXPORT PartitionedHostStateMapBase {
   bool HasPartitionHash() const;
   // Returns true if |partition_hash_| contains a non empty valid hash.
   bool IsPartitionHashValid() const;
-  // Creates a host hash by concatenating first 16 bytes (half of SHA256) from
-  // |k| and first 16 bytes from |partition_hash_|.
+  // Returns first `HashedHost::size() / 2` bytes from |k|.
```
Comment (goodov) [inline net/http/partitioned_host_state_map.h:40]: this comment should go near `GetHalfKey`

File: net/http/partitioned_host_state_map_unittest.cc [hunk 20]
```diff
@@ -16,8 +16,8 @@ namespace net {
...
 
-using PartitionedMap = PartitionedHostStateMap<
-    std::map<std::array<uint8_t, crypto::kSHA256Length>, std::string>>;
+using HostHash = std::array<uint8_t, crypto::kSHA256Length>;
+using PartitionedMap = PartitionedHostStateMap<std::map<HostHash, std::string>>;
 
 std::array<uint8_t, crypto::kSHA256Length> HashHost(
```
Comment (goodov) [inline net/http/partitioned_host_state_map_unittest.cc:22]: `HostHash HashHost(...`

File: chromium_src/chrome/browser/ui/views/bubble/bubble_contents_wrapper.h [hunk 21]
```diff
@@ -24,9 +24,8 @@
                       WindowOpenDisposition disposition,                   \
                       const blink::mojom::WindowFeatures& window_features, \
                       bool user_gesture, bool* was_blocked) override;      \
-  std::vector<int32_t>& GetPopupIds() {                                    \
-    return popup_ids_;                                                     \
-  }                                                                        \
+  const std::vector<int32_t>& popup_ids() { return popup_ids_; }           \
```
Comment (goodov) [inline chromium_src/chrome/browser/ui/views/bubble/bubble_contents_wrapper.h]: `popup_ids() const`

**Review by goodov — APPROVED** [review]
lgtm with few nits.

File: net/http/partitioned_host_state_map.h (hunk 19 above)
Comment (emerick) [reply net/http/partitioned_host_state_map.h]: Oops, fixed in eb96a2aa698459aacbf30e3e2196b95e0093d33b.

File: net/http/partitioned_host_state_map_unittest.cc (hunk 20 above)
Comment (emerick) [reply net/http/partitioned_host_state_map_unittest.cc]: Fixed in eb96a2aa698459aacbf30e3e2196b95e0093d33b.

File: chromium_src/chrome/browser/ui/views/bubble/bubble_contents_wrapper.h (hunk 21 above)
Comment (emerick) [reply chromium_src/chrome/browser/ui/views/bubble/bubble_contents_wrapper.h]: Fixed in 9501de7a282bd7b5f38697159793ee3a856db1eb.

File: chromium_src/extensions/browser/sandboxed_unpacker.cc (hunk 16 above)
Comment (emerick) [reply chromium_src/extensions/browser/sandboxed_unpacker.cc:45]: @yrliou Do you happen to know if we still need to rewrite the manifest to add socket permissions for the IPFS Companion extensions?

File: patches/chrome-browser-buildflags.gni.patch (hunk 17 above)
Comment (emerick) [reply patches/chrome-browser-buildflags.gni.patch]: Fixed in 7996e758b709a434e0808e83e0f6a18f07ea0948.

File: build/rust/Cargo.lock [hunk 22]
```diff
@@ -551,17 +551,17 @@ dependencies = [
 
 [[package]]
 name = "cxx"
-version = "1.0.81"
+version = "1.0.82"
```
[the remark this answers is not in the document]
Comment (rillian) [reply build/rust/Cargo.lock]: Note the main branch has 1.0.68 now. We should keep that version when resolving merging/rebase conflicts.

File: chromium_src/extensions/browser/sandboxed_unpacker.cc (hunk 16 above)
Comment (bridiver) [reply chromium_src/extensions/browser/sandboxed_unpacker.cc:45]: @emerick I also pinged @cypt4

**Review by bridiver — APPROVED** [review]

**Review by iefremov — APPROVED** [review]

File: chromium_src/extensions/browser/sandboxed_unpacker.cc (hunk 16 above)
Comment (emerick) [reply chromium_src/extensions/browser/sandboxed_unpacker.cc]: No longer needed, removed here 1893df15a2a1f4ef4cd3864cbc158341415189a7.

**Pull request merged**